MovieQuiz/Services/QuestionFactory.swift

```diff
@@ -8,85 +8,101 @@
 import Foundation
 
 final class QuestionFactory: QuestionFactoryProtocol {
-    private var questions: [QuizQuestion] = []
+    private let moviesLoader: MoviesLoading
+    private var movies: [MostPopularMovie] = []
+    private var unusedQuestionCount: Int = 0
     
     weak var delegate: QuestionFactoryDelegate?
     
-    init(delegate: QuestionFactoryDelegate?) {
-        self.delegate = delegate
-        questions = loadMockData()
+    init(moviesLoader: MoviesLoading) {
+        self.moviesLoader = moviesLoader
     }
     
     func requestNextQuestion() {
-        guard let index = (0..<questions.count).randomElement() else {
-            delegate?.didRecieveNextQuestion(self, question: nil)
+        guard !movies.isEmpty else {
+            loadData()
             return
         }
-        let question = questions[safe: index]
-        if question != nil {
-            deleteIndexQuestionOrReloadQuestions(index)
+        
+        DispatchQueue.global().async { [weak self] in
+            guard let self = self else { return }
+            let index = (0..<self.unusedQuestionCount).randomElement() ?? 0
+            
+            guard let movie = self.movies[safe: index] else { return }
+            
+            let question = self.convert(model: movie)
+            self.hideIndexQuestionOrUpdateAllQuestions(index)
+            
+            DispatchQueue.main.async { [weak self] in
+                guard let self = self else { return }
+                self.delegate?.didRecieveNextQuestion(self, question: question)
+            }
+        }
+    }
+    
+    func loadData() {
+        moviesLoader.loadMovies { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .success(let mostPopularMovies):
+                var errorMessage = mostPopularMovies.errorMessage
+                if errorMessage.isEmpty, mostPopularMovies.items.isEmpty {
+                    errorMessage = "The data is not loaded!"
+                }
+                if !errorMessage.isEmpty {
+                    DispatchQueue.main.async { [weak self] in
+                        guard let self = self else { return }
+                        self.delegate?.didFailToLoadData(self, with: ServiceError.general(reason: errorMessage))
+                    }
+                    return
+                }
```

Comment:
if с return обычно принято переделывать в guard else. При этом, в блоке else у guard стараются не писать много кода. Я бы эту часть отрефакторинла, а содержимое вынесла в отдельную функцию 

```suggestion
                guard errorMessage.isEmpty else {
                    self.handle(errorMessage)
                    return
                }
```

Reply:
Понял, самому этот код не очень, сначала был guard, смутил объем кода перекинул в  if
Как вот такой вариант: 
`guard checkErrorMessage(mostPopularMovies.errorMessage, mostPopularMovies.items.isEmpty) else { return } `
или я так выпадаю из общего контекста или в сам guard не принято писать функцию?

Reply:
Нет, в guard можно вызывать  функции :)